### DISCUSSION TITLE
Shutters Auto Calibration

### DISCUSSION THREAD

**Discussion by najdanovicivan** [arendst/Tasmota#13688] [Ideas]
Regarding the Shutters and Blinds I have the motors which have mechanical limiters. Once the motor reaches the end position it will automatically stop. If will be great for devices which have means of measuring power consumption to implement some sort of auto calibration feature. Looking at the power consumption to automatically measure the time needed to open and close the shutters.

**Comment by Funkenschuhster** [comment]
On addition to that a semi automatic calibration can be implemented. Like:

1. Enable calibration. --> Software endstop override active
2. Open shutter manually
3. Confirm 100% Open --> Internal stopwatch is set to 0s
4. Close shutter and release button immediately when fully closed. Tasmota measures time during movement
5. Confirm 0% --> ShutterCloseDuration = Stopwatch & Stopwatch = 0s
6. Open shutter and release button immediately when 100% Open --> time is measured by tasmota
7. Confirm 100% --> ShutterOpenDuration = Stopwatch
8. disable calibration. --> sw endstop active

At least a stopwatch which measures relais time would be helpful.

**Reply by stefanbode** [reply]
The reason for autocallibration not part of the driver is the overhead of code. This is only needed once and every byte on the driver is something to think about. This will go away with ESP32. But all ESP8266 must be as lean as possible.

**Comment by JosefRypacek** [comment]
This would be useful for Sonoff dual R3 as it is ESP32 and this functionality is implemented in official firmware.

**Reply by stefanbode** [reply]
Yes this is a great idea I was thinking of for ESP32. It is a little impossible for the small one. I have to look what magic they are doing. I assume they observe the current. Looks an interesting tasks.

**Reply by JosefRypacek** [reply]
Thanks! Yes, it would be based on current. I have found something relevant in ESPHome (not tested). 
https://esphome.io/components/cover/current_based.html

This is basically nice to have feature as everybody capable of flashing Tasmota should be able to set it manually, but we should use capabilities of the device (if available).

**Reply by drrokko** [reply]
keep in mind, there is a "Sonoff DUAL R3 Lite" which has no power measurement, independent of ESP32 or not

**Reply by stefanbode** [reply]
This should be no issue. If there is no power management the function will not execute. I still have no idea because of the crashes.
